mypy/checkexpr.py

```diff
@@ -2360,6 +2360,12 @@ def lookup_definer(typ: Instance, attr_name: str) -> Optional[str]:
             any_type = AnyType(TypeOfAny.from_another_any, source_any=right_type)
             return any_type, any_type
 
+        # If either the LHS or the RHS is Instance and fallbacks to Any, we also return Any
+        if ((isinstance(left_type, Instance) and left_type.type.fallback_to_any) or
+                (isinstance(right_type, Instance) and right_type.type.fallback_to_any)):
+            any_type = AnyType(TypeOfAny.special_form)
+            return any_type, any_type
```

Comment:
Honestly this looks a bit too radical. But I am not a big expert here, @Michael0x2a maybe you can review this (feel free to un-assign if you don't have time)?

Reply:
I think this would be fine only after the below checks for the presence of `__op__`/`__rop__`.  E.g. this should still infer a precise type:

```
from typing import Any
C: Any
class D(C):
    def __add__(self, rhs: float) -> D:
        return self

reveal_type(D() + 0.5)  # N: Revealed type is 'D'
```